### DISCUSSION TITLE
How to define how sympy should print an instance of a class?

### DISCUSSION THREAD

**Discussion by evbernardes** [sympy/sympy#22068] [Q&A]
Taking the example of the Quaternion class, since I've been using it the most, I found the function "_print_Quaternion" inside of printing/latex.py.

How is it called exactly and how do I quickly implement this if I'm making a new class?

Thanks in advance!

**Comment by asmeurer** [comment]
Each printer has a method you can define on the class. For LaTeX it is `_latex`. See https://docs.sympy.org/latest/modules/printing.html#example-of-custom-printing-method.

**Reply by evbernardes** [reply]
That's what I've been doing. For example, I defined a symbolic operation in matrices/expressions/hat.py

There I defined a class called Hat, in _latex.py I defined a function called _print_Hat and it works flawlessly.

On the other hand:

 I created a submodule, let's call it module

Inside this module o created, inside of file sympy/module/base.py a base class that  inherits from Expr, let's call it Base.

Then in file sympy/module/testclass.py I created a class TestClass that inherits from Base and added _print_TestClass in _latex.py but this time it didn't work.

My question is, more precisely, should I add some kind of reference to this module somewhere?

**Reply by ThePauliPrinciple** [reply]
You can define a _latex method in your custom class, as in
```python
class TestClass:
  def _latex(self,printer):
    return "the string you want to be printed"
```
You can use printer._print(object) to print any sympy objects (it will then automatically select an appropriate printing method), for example, if your class is something representing a function, it could print the arguments.
